pkg/inject/sidecar.go

```diff
@@ -0,0 +1,99 @@
+package inject
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/api/core/v1"
+
+	"github.com/jaegertracing/jaeger-operator/pkg/apis/io/v1alpha1"
+	"github.com/jaegertracing/jaeger-operator/pkg/deployment"
+	"github.com/jaegertracing/jaeger-operator/pkg/service"
+)
+
+var (
+	// Annotation is the annotation name to look for when deciding whether or not to inject
+	Annotation = "inject-jaeger-agent"
+)
+
+// Sidecar adds a new container to the deployment, connecting to the given jaeger instance
+func Sidecar(dep *appsv1.Deployment, jaeger *v1alpha1.Jaeger) {
+	deployment.NewAgent(jaeger) // we need some initialization from that, but we don't actually need the agent's instance here
+
+	if jaeger == nil || dep.Annotations[Annotation] != jaeger.Name {
+		logrus.Debugf("Skipping sidecar injection for deployment %v", dep.Name)
+	} else {
+		logrus.Debugf("Injecting sidecar for pod %v", dep.Name)
+		dep.Spec.Template.Spec.Containers = append(dep.Spec.Template.Spec.Containers, container(jaeger))
+	}
+}
+
+// Needed determines whether a pod needs to get a sidecar injected or not
+func Needed(dep *appsv1.Deployment) bool {
+	if dep.Annotations[Annotation] == "" {
+		logrus.Debugf("Not needed, annotation not present for %v", dep.Name)
+		return false
+	}
+
+	// this pod is annotated, it should have a sidecar
+	// but does it already have one?
+	hasSidecar := false
+	for _, container := range dep.Spec.Template.Spec.Containers {
+		if container.Name == "jaeger-agent" { // we don't labels/annotations on containers, so, we rely on its name
+			hasSidecar = true
+		}
+	}
+
+	return !hasSidecar
+}
+
+// Select a suitable Jaeger from the JaegerList for the given Pod, or nil of none is suitable
+func Select(target *appsv1.Deployment, availableJaegerPods *v1alpha1.JaegerList) *v1alpha1.Jaeger {
+	jaegerName := target.Annotations[Annotation]
+	if strings.ToLower(jaegerName) == "true" && len(availableJaegerPods.Items) == 1 {
+		// if there's only *one* jaeger within this namespace, then that's what
+		// we'll use -- otherwise, we should just not inject, as it's not clear which
+		// jaeger instance to use!
+		// first, we make sure we normalize the name:
+		jaeger := &availableJaegerPods.Items[0]
+		target.Annotations[Annotation] = jaeger.Name
```

Comment:
the target is not modified in the following loop

Reply:
This is modified because it might be "true", so, it becomes the Jaeger's name. It's not modified in the following loop because there, the annotation content == jaeger.Name already, so, there's nothing to normalize.